### PR TITLE
ILogException support for logging exceptions with error codes

### PR DESCRIPTION
**Pull request by emgarten**
Adding ``ILogException.AsLogMessage`` for Exceptions. This allows logging the exception to an ILogger with complete metadata.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/ExceptionUtilities.cs [hunk 1]
```diff
@@ -14,20 +14,42 @@ namespace NuGet.Common
...
-                logger.LogError(ex.ToString());
+                // Log the log message itself.
+                var logMessage = logExceptionMessage.AsLogMessage();
+                logger.Log(logMessage);
             }
             else
             {
-                logger.LogError(ExceptionUtilities.DisplayMessage(ex));
+                // Create a string from the exception.
+                logger.Log(LogMessage.CreateError(NuGetLogCode.NU1000, DisplayMessage(unwrappedException)));
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Common/ExceptionUtilities.cs]: we should consider improving this in the future. Else everything will throw NU1000 with an unexpected exception.

File: src/NuGet.Core/NuGet.Packaging/MinClientVersionUtility.cs [hunk 2]
```diff
@@ -77,6 +85,12 @@ public static bool IsMinClientVersionCompatible(NuGetVersion packageMinClientVer
         /// </summary>
         public static NuGetVersion GetNuGetClientVersion()
         {
+            // Use the override if it was given.
+            if (ClientVersionOverride != null)
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/MinClientVersionUtility.cs]: move this into next if block.

**Review by mishra14 — APPROVED** [review]
Small changes. then nugget it.

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs [hunk 3]
```diff
@@ -81,8 +81,7 @@ public static int MainCore(string workingDirectory, string[] args)
 
             var console = new Console();
             var fileSystem = new CoreV2.NuGet.PhysicalFileSystem(workingDirectory);
-
-            Func<Exception, string> getErrorMessage = ExceptionUtilities.DisplayMessage;
+            var logStackAsError = console.Verbosity == Verbosity.Detailed;
```
Comment (rohit21agrawal) [inline src/NuGet.Clients/NuGet.CommandLine/Program.cs:84]: do you also need to check for diagnostic verbosity?

File: src/NuGet.Core/NuGet.Common/ExceptionUtilities.cs [hunk 4]
```diff
@@ -14,20 +14,42 @@ namespace NuGet.Common
     /// </summary>
     public static class ExceptionUtilities
     {
-        public static void HandleException(Exception ex, ILogger logger)
+        /// <summary>
+        /// Log an exception to an ILogger.
+        /// This will log using NU1000 if the exception does not contain a code.
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Common/ExceptionUtilities.cs:19]: does it have to contain a code necessarily?

Comment (emgarten) [reply src/NuGet.Core/NuGet.Common/ExceptionUtilities.cs:19]: No, I'll remove this and add an undefined code so top level these can still be printed without a code.

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs (hunk 3 above)
Comment (emgarten) [reply src/NuGet.Clients/NuGet.CommandLine/Program.cs:84]: that doesn't exist for nuget.exe, there are just 3

**Pull request closed without merging**